pkg/gds/members.go

```diff
@@ -97,6 +103,109 @@ func (s *Members) Shutdown() (err error) {
+	if in.PageSize == 0 {
+		in.PageSize = defaultPageSize
+	}
+
+	// If a page cursor is provided, load it - otherwise create a cursor for iteration
+	cursor := &models.PageCursor{}
+	if in.PageToken != "" {
+		if err = cursor.Load(in.PageToken); err != nil {
+			log.Warn().Err(err).Msg("invalid page token on list request")
+			return nil, status.Error(codes.InvalidArgument, "invalid page token")
+		}
+
+		// Validate the request has not changed
+		if cursor.PageSize != in.PageSize {
+			log.Debug().Int32("cursor", cursor.PageSize).Int32("opts", in.PageSize).Msg("invalid members list request: mismatched page size")
+			return nil, status.Error(codes.InvalidArgument, "page size cannot change between requests")
+		}
+
+	} else {
+		// Update the cursor with the input request
+		cursor.PageSize = in.PageSize
+	}
+
+	// Create response
+	out = &api.ListReply{
+		Vasps: make([]*api.VASPMember, 0, cursor.PageSize),
+	}
+
+	// Create the VASPs iterator to begin collecting validated VASPs data
+	iter := s.db.ListVASPs()
+	defer iter.Release()
+
+	// If necessary, seek to the next key specified by the cursor.
+	if cursor.NextVasp != "" {
+		// If iter.Seek returns false (e.g. seek did not find the specified key) then
+		// iter.Next() should also return false, so it isn't necessary to check the return.
+		// NOTE: next key must be deleted after it's used for seeking so that the last
+		// page doesn't retain the old key and loop forever.
+		iter.Seek(cursor.NextVasp)
+		cursor.NextVasp = ""
+
+		// Because we're going to be calling Next, we need to back up one key to ensure
+		// that we start on the right key in the for loop.
+		iter.Prev()
+	}
+
+	// Iterate over VASPs, collecting the one we're looking for.
+	for iter.Next() {
+		// Check if we're done iterating - if so and there is next data, there is
+		// another page, so create the page token to return it.
+		if len(out.Vasps) == int(cursor.PageSize) {
+			cursor.NextVasp = iter.Id()
+			break
+		}
+
+		// Collect the VASP from the iterator
+		vasp := iter.VASP()
+
+		// Skip any VASPs that are not verified yet
+		if vasp.VerificationStatus != pb.VerificationState_VERIFIED {
+			continue
+		}
+
+		// Build the directory information to return to the user
+		info := &api.VASPMember{
+			Id:                  vasp.Id,
+			RegisteredDirectory: vasp.RegisteredDirectory,
+			CommonName:          vasp.CommonName,
+			Endpoint:            vasp.TrisaEndpoint,
+			Website:             vasp.Website,
+			BusinessCategory:    vasp.BusinessCategory,
+			VaspCategories:      vasp.VaspCategories,
+			VerifiedOn:          vasp.VerifiedOn,
+		}
+
+		// Add other information to the VASP
+		if info.Name, err = vasp.Name(); err != nil {
+			log.Error().Err(err).Str("vasp_id", vasp.Id).Msg("could not retrieve VASP name from record")
+		}
+
+		if vasp.Entity != nil {
+			info.Country = vasp.Entity.CountryOfRegistration
+		}
+
+		out.Vasps = append(out.Vasps, info)
+	}
+
+	if err = iter.Error(); err != nil {
```

Comment:
Does the call to `iter.Error` here just poll the iterator for any errors it collected during the `iter.Next` loop?

Reply:
Yes, if there is an error during processing `iter.Next()` will return false breaking the loop.